lib/worker/output.go

```diff
@@ -0,0 +1,219 @@
+	go o.handleOutput()
+	go o.handleBroadcast()
+	go o.handleErrors()
+	return o, nil
+}
+
+// Stream returns a channel to receive output up to the
+// present time and stream real time. The channel closes when
+// there the process the output is coming from is finished.
+// Any errors encountered during output reading are sent through
+// the errors channel.
+func (o *OutputHandler) Stream() (chan ProcessOutputEntry, chan error) {
+	output := make(chan ProcessOutputEntry)
+	errChan := make(chan error)
+
+	go func() {
+		if o.bufferLen() > 0 {
+			output <- ProcessOutputEntry{Content: o.combinedBuffer}
+		}
+		select {
+		case <-o.done:
+			close(output)
+		default:
+			o.addListener(&Listener{
+				outputChan: output,
+				errorChan:  errChan,
+			})
+		}
+	}()
+
+	return output, errChan
+}
+
+func (o *OutputHandler) addListener(newListener *Listener) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.listeners = append(o.listeners, newListener)
+}
+
+func (o *OutputHandler) updateCombinedBuffer(b []byte) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	o.combinedBuffer = append(o.combinedBuffer, b...)
+}
+
+func (o *OutputHandler) bufferLen() int {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return len(o.combinedBuffer)
+}
+
+func (o *OutputHandler) iterListeners(routine func(*Listener)) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	for _, l := range o.listeners {
+		routine(l)
+	}
+}
+
+// handleBroadcast forwads new messages to the listeners and
+// closes listener channels when done signal is received to
+// signal end of stream.
+func (o *OutputHandler) handleBroadcast() {
+L:
+	for {
+		select {
+		case msg := <-o.messages:
+			o.iterListeners(func(l *Listener) {
+				l.outputChan <- ProcessOutputEntry{Content: msg}
+			})
+		case <-o.done:
+			o.iterListeners(func(l *Listener) {
+				close(l.outputChan)
+			})
+			break L
+		}
+	}
+}
+
+// handleErrors forwards errors to the listeners
+func (o *OutputHandler) handleErrors() {
+L:
+	for {
+		select {
+		case err := <-o.errors:
+			o.iterListeners(func(l *Listener) {
+				l.errorChan <- err
+			})
+			break L
+		}
+	}
+}
+
+// handleOutput manages the goroutines that consume
+// the reader to send to buffer and listeners. It
+// closes the done channel done.
+func (o *OutputHandler) handleOutput() error {
+	wg := sync.WaitGroup{}
+
+	readersLen := len(o.readers)
+	wg.Add(readersLen)
+
+	// TODO: ideally, we would have a mechanism to halt
+	// all goroutines in this OutputHandler if there
+	// is an error. Like a quit channel.
+	for _, reader := range o.readers {
+		go func(r io.Reader) {
+			o.bufferAndForwardChunks(r)
+			wg.Done()
+		}(reader)
+	}
+
+	go func() {
+		wg.Wait()
+		// a closed channel is always ready to receive
+		close(o.done)
+	}()
+
+	return nil
+}
+
+// bufferAndForwardChunks reads the reader by 76 byte chunks
+// or it will stop at EOF. It buffers each chunk and forwards
+// it to the messages channel.
+func (o *OutputHandler) bufferAndForwardChunks(reader io.Reader) {
+	buf := make([]byte, 0, 76)
```

Comment:
I'm reasonably convinced that `io.ReadFull()` will handle the case where the process only ever emits 50 bytes. 

What if the process only emits 50 bytes and then nothing else for an hour - this will block for that entire time yeah? What are the tradeoffs with doing that?